rustls/src/error.rs

```diff
@@ -99,6 +99,92 @@ pub enum Error {
     BadMaxFragmentSize,
 }
 
+#[non_exhaustive]
+#[allow(missing_docs)]
+#[derive(Debug, PartialEq, Clone)]
+/// The set of cases where we failed to make a connection because we thought
+/// the peer was misbehaving.
+///
+/// This is `non_exhaustive`: we might add or stop using items here in minor
+/// versions.  We also don't document what they mean.  Generally a user of
+/// rustls shouldn't vary its behaviour on these error codes, and there is
+/// nothing it can do to improve matters.
+///
+/// Please file a bug against rustls if you see `Error::PeerMisbehaved` in
+/// the wild.
+pub enum PeerMisbehaved {
+    AttemptedDowngradeToTls12WhenTls13IsSupported,
+    BadCertChainExtensions,
+    ClientHelloSniMustContainOneHostName,
+    DisallowedEncryptedExtension,
+    DuplicateClientHelloExtensions,
+    DuplicateEncryptedExtensions,
+    DuplicateHelloRetryRequestExtensions,
+    DuplicateNewSessionTicketExtensions,
+    DuplicateServerHelloExtensions,
+    DuplicateServerNameTypes,
+    EarlyDataAttemptedInSecondClientHello,
+    EarlyDataExtensionWithoutResumption,
+    EarlyDataOfferedWithVariedCipherSuite,
+    HandshakeHashVariedAfterRetry,
+    IllegalHelloRetryRequestWithEmptyCookie,
+    IllegalHelloRetryRequestWithNoChanges,
+    IllegalHelloRetryRequestWithOfferedGroup,
+    IllegalHelloRetryRequestWithUnofferedCipherSuite,
+    IllegalHelloRetryRequestWithUnofferedNamedGroup,
+    IllegalHelloRetryRequestWithUnsupportedVersion,
+    IllegalMiddleboxChangeCipherSpec,
+    IllegalTlsInnerPlaintext,
+    IncorrectBinder,
+    InvalidMaxEarlyDataSize,
+    InvalidKeyShare,
+    InvalidSctList,
+    KeyEpochWithPendingFragment,
+    KeyUpdateReceivedInQuicConnection,
+    MessageInterleavedWithHandshakeMessage,
+    MissingBinderInPskExtension,
+    MissingKeyShare,
+    MissingQuicTransportParameters,
+    OfferedDuplicateKeyShares,
+    OfferedEarlyDataWithOldProtocolVersion,
+    OfferedEmptyAlpnProtocol,
+    OfferedIncorrectCompressions,
+    PskExtensionMustBeLast,
+    PskExtensionWithMismatchedIdsAndBinders,
+    RefusedToFollowHelloRetryRequest,
+    RejectedEarlyDataInterleavedWithHandshakeMessage,
+    ResumptionAttemptedWithVariedEms,
```

Comment:
Should we expand `Ems` in these errors?

Reply:
> I think `kx` or `ems` will be confusing for many of our users.

I don't think our users will ever see these errors, and if they do they should definitely be for our eyes as they're not really user-actionable -- and that's why the docs for `PeerMisbehaved` asks for a bug to be filed. `PeerIncompatible` is arguably more user-visible and might even be user-actionable, hence why `ServerTlsVersionIsDisabledByOurConfig` is a separate error.

Also, we use "kx group" elsewhere in the public API, like `SupportedKxGroup` etc.

>`Sni` is a well-known extension name but in context expanding to `ServerName` seems like an improvement, similar for `Alpn` vs `ApplicationProtocol`.

Agree on both these though.

Reply:
(I would probably be in favor of renaming `SupportedKxGroup` to `SupportedKeyExchangeGroup`, FWIW.)

Honestly, even for me `Ems` is a little out there still -- not sure how commonly that acronym is used.